[Riako] 사다리 구현 4 단계 - 리팩토링 2

naneun opened the pull request:
## PR 인사
> 안녕하세요 `호눅스`! 😀
저번 단계에서 리뷰해주신 내용을 토대로 전반적으로 수정해봤습니다!
이와 더불어 불필요한 클래스들을 대다수 삭제하고 꼭 필요한 로직만 담을 수 있도록 노력했습니다.
이번 단계에서도 좋은 조언 부탁드리겠습니다. ✨

## PR 목적
  >- 사다리 구현 4 단계 - 리팩토링 2 작업 반영

## 프로그래밍 요구사항
  >- [x] 메소드의 크기가 최대 10라인을 넘지 않도록 구현한다.
  >- [x] method가 한 가지 일만 하도록 최대한 작게 만들어라.
  >- [x] indent(인덴트, 들여쓰기) depth를 2단계에서 1단계로 줄여라.
  >- [x] depth의 경우 if문을 사용하는 경우 1단계의 depth가 증가한다. if문 안에 while문을 사용한다면 depth가 2단계가 된다.
  >- [x] else를 사용하지 마라.
  >- [x] 배열 대신 ArrayList와 Generic을 활용해 구현한다.
  >- [x] 로직을 구현하는 코드에 단위 테스트가 존재해야 한다. 단, UI 처리 로직(System.in, System.out)은 테스트에서 제외한다.
  >- [x] 각각의 역할에 맞도록 패키지를 분리하고 접근 제어자를 적절히 사용하도록 리팩토링한다.

## 결과화면
  >![SmartSelectImage_2022-02-18-01-37-01](https://user-images.githubusercontent.com/47964708/154527642-da50ebd9-4603-4b22-8a0e-85d8527cfe72.png) 

## 특이사항
  >- 저번 단계에서와는 다르게 평소 프로그래밍 습관으로 코드를 작성해보았습니다.

## 고민사항
  >- 클래스를 최대한 많이 쪼개는 것이 객체 지향 프로그래밍에서 좋은 방향이 맞는 지 궁금합니다.
  >- commit 메시지에 이모지를 써봤는데 현업에서도 많이 사용하나요? 보기도 깔끔하고 귀여워서 이것만 쓸 것 같습니다.

## 마무리
  > 감사합니다 🤗

naneun committed:
🚜 미션3에서 리뷰 받은 내용을 바탕으로 구조 개선

naneun committed:
☔️ Line 클래스 테스트 코드 작성

- isValidTrue: 해당 위치에 가로선을 놓을 수 있다면 true
- isValidFalse: 해당 위치에 가로선을 놓을 수 없다면 true

naneun committed:
🔨 Line 클래스 메서드 접근자 수정 (public -> default)

- 같은 패키지에서 Ladder 클래스가 Line 메서드를 사용하므로 접근자를 수정했습니다.

naneun committed:
📝 테스트 코드 @DisplayName 작성

naneun committed:
📚 사다리 구현 4단계 - 리팩토링 2 관련 내용 문서 작성

naneun committed:
📚 실행 결과화면 첨부

naneun committed:
📝 appendElement -> addElement 메서드 명 변경

naneun committed:
☔️ 플레이어 이름 출력 테스트

- 글자 수에 상관 없이 사다리 간격의 길이에 맞게 이름이 출력되는 지 테스트
- 글자 수가 5개 이상일 경우 .. 을 붙였습니다.

naneun committed:
☔️ addElement 메서드 경계값 테스트 추가

naneun committed:
📝 InputViewTest -> InputUtilsTest 테스트 클래스 명 변경

wheejuni commented on src/main/java/application/domain/LadderGame.java:
```diff
@@ -0,0 +1,19 @@
+import application.domain.ladder.Ladder;
+import application.domain.player.PlayerList;
+
+public class LadderGame {
```
단순히 사다리와 플레이어 리스트를 보관만 할 뿐 맡은 역할은 없군요...
조금 아쉬운 설계인데 그러면 실제 플레이는 어디서 일어나야 할까요? `LadderGame` 이라는 클래스명과 다소 맞지 않는 설계는 아니려나요.

naneun replied:
현재 5단계에서는 사다리 게임을 진행하는 기능이 LadderGame 에 추가되었는데 이를 고려하면 괜찮은 클래스 명일까요? 실제 플레이가 일어나는 곳도 LadderGame 이라 생각해서 이렇게 설계했었는데 리뷰해주신 내용을 보니 좀 더 고민을 해봐야겠습니다!

wheejuni commented on src/main/java/application/domain/element/LadderElement.java:
```diff
@@ -0,0 +1,17 @@
+package application.domain.element;
+
+public enum LadderElement {
+    STEP("-----")
+    , EMPTY("     ")
+    , RAIL("|");
```
콤마는 한 줄의 끝에 작성해주는 것이 컨벤션입니다.

```suggestion
    STEP("-----"),
    EMPTY("          "),
    RAIL("|");

wheejuni commented on src/main/java/application/domain/element/LadderElement.java:
```diff
@@ -0,0 +1,17 @@
+
+    private String str;
+
+    public String toString() {
+        return str;
+    }
```
`.toString()` 이 아닙니다. 이 메소드명은 `java.lang.Object` 에 작성되어 있는 것으로만 사용해 주시고요.
이 메소드가 갖는 역할을 위한 더 좋은 이름을 구상해주세요. 앞으로 메소드명을 설계하실 때도 `java.lang.Object` 에 이미 존재하는 이름들은 피해주시는 게 좋겠습니다.

wheejuni commented on src/main/java/application/domain/ladder/Line.java:
```diff
@@ -0,0 +1,34 @@
+public class Line {
+    private List<LadderElement> line;
+
+    Line() {
+        this.line = new ArrayList<>();
+    }
+
+    void addElement(LadderElement e) {
+        line.add(e);
+    }
+
+    boolean isValid(int col) {
+        return col == 0 || line.get(col - 1) != STEP;
+    }
```
무슨 이유에선지 여기는 접근 제어자가 다 누락되어 있군요.

naneun replied:
접근 제어자를 사용해보라는 요구 사항이 있어 자주 사용하지 않았던 default 접근자를 사용해 같은 패키지 내에 있는 Ladder 클래스만이 Line 클래스의 기능을 사용할 수 있도록 해봤습니다!

wheejuni commented on src/main/java/application/domain/player/Player.java:
```diff
@@ -1,14 +1,24 @@
     @Override
     public String toString() {
-        return name;
+        if (name.length() > Ladder.INTERVAL) {
+            name = name.substring(0, Ladder.INTERVAL - 2).concat("..");
+        }
+        return String.format("%-" + (Ladder.INTERVAL + 1) + "s", name);
```
`.toString()` 에 들어가기엔 너무 복잡한 로직인 것 같습니다!
이렇게 자세한 정보를 `.toString()` 으로 표현할 필요 없고요, 애플리케이션 로직이나 뷰 단에서 사용할 문자열을 여기에서 생성하시면 안됩니다.
그런 필요가 있다면 별도의 메소드를 설계해주세요.

naneun replied:
별도의 메소드로 정의했습니다!

wheejuni commented on src/main/java/application/domain/player/Player.java:
```diff
@@ -1,14 +1,24 @@
         this.name = name;
     }
 
+    public String getName() {
+        return name;
+    }
```
로직에 따라 사용자 이름을 표시해줘야 할텐데 게터 꼭 필요한지요?

naneun replied:
제 테스트 코드에서 실제 이름의 길이를 사용하여 테스트하는 부분이 있어서 넣었었는데 조금 더 생각해보겠습니다!

naneun replied:
제거했습니다!

wheejuni commented on src/main/java/application/domain/player/PlayerList.java:
```diff
@@ -0,0 +1,22 @@
+
+    public PlayerList(List<Player> players) {
+        this.players = players;
+        this.totalNum = players.size();
```
어차피 `.size()` 로 값을 넣는데, 별도의 인스턴스 변수가 필요한 이유가 궁금해요.

naneun replied:
말씀해주신 내용을 기반으로 생각해봤는데 메서드 내에서 지역변수로 .size() 를 담아 그때그때 사용하는 편이 더 좋을 것 같습니다! 감사합니다 🎉

naneun replied:
말씀해주신 내용 보고 불필요한 부분을 없앨 수 있었습니다! 감사합니다! 🎉

wheejuni reviewed (changes requested):
대단히 수고 많으셨습니다 👍 
코멘트 참고 부탁드릴게요!

naneun committed:
🎨 wheejuni 리뷰어님 리뷰 내용 반영

- ',' 콤마 한 줄 끝에 작성하는 컨벤션 적용
- toString() 에 작성된 로직을 별도의 메소드를 만들어 이동
- 불필요한 getter, 인스턴스 변수 삭제

naneun commented:
안녕하세요 wheejuni 리뷰어 님! 리뷰해주신 내용들을 프로젝트에 반영했습니다.
리뷰해주신 내용 중 질문드릴 게 하나 있습니다!

> 단순히 사다리와 플레이어 리스트를 보관만 할 뿐 맡은 역할은 없군요... 조금 아쉬운 설계인데 그러면 실제 플레이는 어디서 일어나야 할까요? `LadderGame` 이라는 클래스명과 다소 맞지 않는 설계는 아니려나요.

현재 4단계까지는 플레이의 기능이 존재하지 않아 LadderGame 클래스는 단순 출력 기능만을 지녔습니다. 리뷰해주신 내용 중 toString 메소드에 출력 로직을 담으면 안된다는 것을 반영하여 화면에 출력하기 위한 메소드를 따로 작성하였습니다. (+ Ladder 클래스에 생성자에서 사다리 모양을 생성하는 기능을 따로 분리하여 LadderGame 에서 이를 호출할 수 있도록 했습니다!) 이를 기반으로 5단계에서는 플레이하는 메소드를 LadderGame 에 추가하고자하는데 이를 고려하더라도 LadderGame 이라는 클래스 명이 어울리지 않는 것인지 여쭤봅니다! 조언 부탁드리겠습니다 🤗

리뷰 감사드립니다 😀 🎉

naneun committed:
🎨 생성자에서 사다리 생성하는 로직 분리

naneun committed:
🎨 (Application) run 메소드 로직 분리

naneun committed:
🎨 (Application) 메서드 명 변경

- init -> initGame, start -> startGame, end -> endGame

naneun committed:
🎨 간단한 코드 정리

naneun committed:
🎨 문맥에 맞게 ladderGame.start() 위치 변경

naneun committed:
🎨 createLine 메소드의 일부 로직을  createElement 메소드로 분리

honux77 commented on README.md:
````diff
@@ -177,4 +177,68 @@
+    ```bash
+    📘class 📗enum
+  
+    ├──📁main
+    │  ├──📁java
+    │  │  ├──📁application
+    │  │  │  ├──📁domain
+    │  │  │  │  ├──📁element
+    │  │  │  │  │  └──📗LadderElement.java
+    │  │  │  │  ├──📁ladder
+    │  │  │  │  │  ├──📘Ladder.java
+    │  │  │  │  │  └──📘Line.java
+    │  │  │  │  ├──📁player
+    │  │  │  │  │  ├──📘Player.java
+    │  │  │  │  │  └──📘PlayerList.java
+    │  │  │  │  └──📘LadderGame.java
+    │  │  │  ├──📁message
+    │  │  │  │  └──📘GameMessage.java  
+    │  │  │  ├──📁util
+    │  │  │  │  ├──📘InputUtils.java  
+    │  │  │  │  └──📘RandomUtils.java
+    │  │  │  ├──📁view
+    │  │  │  │  ├──📘InputView.java 
+    │  │  │  │  └──📘OutputView.java
+    │  │  │  └──📘Application.java
+    └──└──└──📘Main.java
````
헛 이거 직접 그린건가요? 아님 표시하는 다른 방법이 있나요????

naneun replied:
.. 아쉽게도 수작업입니다! 😥

honux77 commented:
제가 머지할게요 ^^ 수고하셨습니다. LadderGame은 이름이 이상한게 아니고 맡은 책임과 역할이 다소 불분명한 점 때문에 코멘트가 나온 것 같네요. Application 이 할 일을 줄여서 LadderGame이 맡으면 괜찮지 않을까 하는 생각이 듭니다.

naneun commented:
> 제가 머지할게요 ^^ 수고하셨습니다. LadderGame은 이름이 이상한게 아니고 맡은 책임과 역할이 다소 불분명한 점 때문에 코멘트가 나온 것 같네요. Application 이 할 일을 줄여서 LadderGame이 맡으면 괜찮지 않을까 하는 생각이 듭니다.

머지 감사합니다! ㅠㅠ 5단계 올리면 리뷰해주시나요?! 일단 올려보겠습니다 😀
